### PR TITLE
Add auth for workflows

### PR DESCRIPTION
**Pull request by loldot**
- [x] New feature
- [ ] Bug fix
- [ ] High impact

**Description of work:**

**Testing:**
- [x] Can be tested
- [x] Automatic tests created / updated
- [x] Local tests are passing

**Checklist:**
- [x] Considered automated tests
- [x] Considered updating specification / documentation
- [x] Considered work items 
- [x] Considered security
- [x] Performed developer testing
- [x] Checklist finalized / ready for review

### CODE REVIEW

File: src/backend/api/Fusion.Resources.Logic/Requests/CanApproveStepHandlerBase.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+using System.Threading.Tasks;
+
+namespace Fusion.Resources.Logic.Requests
+{
+    public record WorkflowAccessKey(string Subtype, string CurrentStep)
+    {
+        public static implicit operator WorkflowAccessKey((string, string) tuple)
+            => new WorkflowAccessKey(tuple.Item1, tuple.Item2);
+    }
+    public record WorkflowAccess()
```
Comment (HansDahle) [inline src/backend/api/Fusion.Resources.Logic/Requests/CanApproveStepHandlerBase.cs:18]: Refactor to seperate file?

File: src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/CanApproveStepHandler.cs [hunk 2]
```diff
@@ -1,42 +1,79 @@
...
-using Fusion.Resources.Database;
 
 namespace Fusion.Resources.Logic.Commands
 {
     public partial class ResourceAllocationRequest
     {
-        public class CanApproveStepHandler : INotificationHandler<CanApproveStep>
+        public class CanApproveStepHandler : CanApproveStepHandlerBase, INotificationHandler<CanApproveStep>
         {
+            private static readonly Dictionary<WorkflowAccessKey, WorkflowAccess> AccessTable = new Dictionary<WorkflowAccessKey, WorkflowAccess>
```
Comment (HansDahle) [inline src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/CanApproveStepHandler.cs:19]: Any chance we could move this into the workflow definition? I guess we can add that to a refactor task.. Maybe align with the fluent structure used there.

File: src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/CanApproveStepHandler.cs [hunk 3]
```diff
@@ -1,42 +1,79 @@
...
+                    .FirstAsync(r => r.Id == notification.RequestId, cancellationToken: cancellationToken);
 
-                //var initiatedBy = await dbContext.Persons.FirstAsync(p => p.Id == notification.InitiatedByDbPersonId);
-                //var request = await dbContext.ResourceAllocationRequests.FirstAsync(r => r.Id == notification.RequestId);
-
-                switch (notification.CurrentStepId)
-                {
-                    case "proposed":
+                var initiator = httpContextAccessor?.HttpContext?.User;
+                if (initiator is null) throw new UnauthorizedWorkflowException();
```
Comment (HansDahle) [inline src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/CanApproveStepHandler.cs:72]: Should always include a message when throwing the same exception in different circumstances.

File: src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/JointVentureAllocationRequestStarted.cs [hunk 4]
```diff
@@ -56,10 +59,11 @@ private void ValidateWorkflow(DbResourceAllocationRequest request)
                             s.AddFailure("proposedPerson", "Must provide a person to be assigned the position"));
                 }
 
-                private ValueTask<bool> AssignRequestToResourceOwnerAsync()
+                private async ValueTask<bool> AssignRequestToResourceOwnerAsync(DbResourceAllocationRequest request)
                 {
-
-                    return new ValueTask<bool>(false);
+                    var department = await router.RouteAsync(request, CancellationToken.None);
```
Comment (HansDahle) [inline src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/JointVentureAllocationRequestStarted.cs:64]: Should not update this. Add this empty default failing method as a placeholder as the step was sort of unclarified from the spec.. But now it is more clear that the JV workflows should never have an approval step. 

But fixing the JV workflow should be done in a separate pr anyways..

File: src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/ResourceOwner/CanApproveResourceOwnerStepHandler.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            public async Task Handle(CanApproveStep notification, CancellationToken cancellationToken)
+            {
+                if (notification.Type != DbInternalRequestType.ResourceOwnerChange) return;
+
+                var request = await dbContext.ResourceAllocationRequests
+                    .Include(p => p.Project)
+                    .FirstAsync(r => r.Id == notification.RequestId, cancellationToken: cancellationToken);
+
+                var initiator = httpContextAccessor?.HttpContext?.User;
+                if (initiator is null) throw new UnauthorizedWorkflowException();
```
Comment (HansDahle) [inline src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/ResourceOwner/CanApproveResourceOwnerStepHandler.cs:61]: Message

File: src/backend/api/Fusion.Resources.Logic/Requests/UnauthorizedWorkflowException.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Runtime.Serialization;
+
+namespace Fusion.Resources.Logic.Requests
+{
+    [Serializable]
+    public class UnauthorizedWorkflowException : Exception
```
Comment (HansDahle) [inline src/backend/api/Fusion.Resources.Logic/Requests/UnauthorizedWorkflowException.cs:7]: Any point including all these constructors when they are never used? Is this som resharper magic template?

File: src/backend/api/Fusion.Resources.Logic/Workflows/AllocationJointVentureWorkflowV1.cs [hunk 7]
```diff
@@ -69,6 +69,7 @@ public WorkflowStep SkipApproval()
             switch (current.Id)
             {
                 case CREATED:
```
Comment (HansDahle) [inline src/backend/api/Fusion.Resources.Logic/Workflows/AllocationJointVentureWorkflowV1.cs:71]: Is this a fix for the workflow or is it regarding auth?

File: src/backend/api/Fusion.Resources.Logic/Requests/CanApproveStepHandlerBase.cs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            {
+                var result = await authorizationService.AuthorizeAsync(initiator, request, OrgProjectAccessRequirement.OrgWrite(request.Project.OrgProjectId));
+                isAllowed |= result.Succeeded;
+            }
+
+            isAllowed |= initiator.IsApplicationUser();
+            isAllowed |= initiator.IsInRole("Fusion.Resources.FullControl");
+            isAllowed |= initiator.IsInRole("Fusion.Resources.Internal.FullControl");
+
+            if (!isAllowed) throw new UnauthorizedWorkflowException();
```
Comment (HansDahle) [inline src/backend/api/Fusion.Resources.Logic/Requests/CanApproveStepHandlerBase.cs:120]: Should include a message with which steps have been evaluated and the different outcomes? 
I think an informative message to the user could save us from much support requests. 

Should also add more telemetry here.. 
In the api we generate a payload based on the evaluated requirements. Maybe we could to that here as well. We should generate a similar payload and trace to AI, that way we can see whats going on _when_ we get questions.

File: src/backend/api/Fusion.Resources.Logic/Requests/CanApproveStepHandlerBase.cs [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+        public CanApproveStepHandlerBase(IAuthorizationService authorizationService)
+        {
+            this.authorizationService = authorizationService;
+        }
+        protected async Task CheckAccess(DbResourceAllocationRequest request, WorkflowAccess row, ClaimsPrincipal initiator)
+        {
+            bool isAllowed = false;
+
+            if (!string.IsNullOrEmpty(request.AssignedDepartment))
```
Comment (HansDahle) [inline src/backend/api/Fusion.Resources.Logic/Requests/CanApproveStepHandlerBase.cs:47]: Could we use the fluent pattern here so we use the `if`'s to build the rule, then only evaluate once.. That way we only evaluate to the first success..

**Review by HansDahle — CHANGES_REQUESTED** [review]

File: src/backend/api/Fusion.Resources.Logic/Requests/UnauthorizedWorkflowException.cs (hunk 6 above)
Comment (loldot) [reply src/backend/api/Fusion.Resources.Logic/Requests/UnauthorizedWorkflowException.cs:7]: It's from the default VS functionality when generating a class which ends with Exception. I guess the template is based on this advise: https://docs.microsoft.com/en-us/dotnet/standard/exceptions/best-practices-for-exceptions#include-three-constructors-in-custom-exception-classes

File: src/backend/api/Fusion.Resources.Logic/Workflows/AllocationJointVentureWorkflowV1.cs (hunk 7 above)
Comment (loldot) [reply src/backend/api/Fusion.Resources.Logic/Workflows/AllocationJointVentureWorkflowV1.cs:71]: It's related to the changes with request router.If there's never a need to approve joint venture requests, we should update the security matrix and revert this.

File: src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/CanApproveStepHandler.cs (hunk 2 above)
Comment (loldot) [reply src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/CanApproveStepHandler.cs:19]: I'll refactor this in another branch.

**Review by HansDahle — APPROVED** [review]
Nice work.. Some warnings on nullability in ApiReportsTo that should be fixed. Should maybe also refactor to separate file?

I think we need to align the JV request with the enterprise request. But in another pr..

**Pull request merged**